(144) 'Tell us more' page: customisable help text

dgmstuart opened the pull request:
- Use a 'desc:' key with the name of a partial
- Describe form is required if no SOR code
  - GOTCHA: clear your session if you have ever sent a request with an SOR code: this will be hanging around in your session cookie
- Removed the parts of the Question code which we didn't end up using

dgmstuart committed:
(refactor) SelectedAnswerStore into app controller

This is used in almost every controller

This is the approach used in the gov.uk Verify app:

https://github.com/alphagov/verify-frontend/blob/verify_frontend_release_342/app/controllers/application_controller.rb#L90-L92

Memoised so that it's only loaded once per request. This probably means
that controller actions can't set an answer, and read out that answer in
the same request, but that would probably not be a good use of
SelectedAnswerStore anyway.

dgmstuart committed:
(refactor) Extract describe repair text to partial

This will allow us to add in custom text later for specific leaf nodes
of the diagnosis tree.

Both views are identical except for the form which is used (passed in by
the controller) and the help text (which we're extracting to partials)

dgmstuart committed:
(refactor) Remove 'describe' from question keys

Currently there are two kinds of key which start with 'describe' - these
lead to a series of options, the other ones lead to 'describe your
repair' pages.

dgmstuart committed:
(spec) Question#redirect_path

The normal behaviour is that it skips straight to the address search
page

dgmstuart committed:
(refactor) Remove top-level 'next' on questions

This was originally intended to support having free-text questions in
the middle of the repair process, which would then need to specify the
next step. We ended up going down a different route, so this
functionality is no longer required.

dgmstuart committed:
(spec) Default Question#redirect_path_for_answer

By default this skips to the describe repair page

dgmstuart committed:
(feature) special describe pages use partials

At the end of some of the diagnosis paths we want to have some specific
content on the required 'describe repair' page so that we can get the
resident to provide information which will be useful in diagnosing their
problem.

Previously our approach only allowed for changing the title. Now we
allow arbitrary html by using partials, specified with a ':desc' key in
the questions.yml

dgmstuart committed:
(feature) Describe form is required if no SOR code

Rather than using two different controllers, make the decision of
whether to make a repair form required or not based on the SOR code:

  - if we have an SOR code, we don't need any additional info, so make
    this optional
  - if we don't have an SOR code, this form is the only way we have of
    helping to diagnose the repair, so it should be required

Note that also previously we weren't saving the description in the
describe_repair_controller! That's an example of how these things can
get out of sync. Having one controller should help avoid that sort of
thing.

Note that we're now checking whether the DescribeRepairForm is valid,
even though it currently has no validations. That's fine: hopefully at
some point it will have some in any case - e.g. field length.

- Use a facade to hide all this from the controller
- In the facade we need to memoize the form: if we don't then each call
  to '#form' creates a new form instance, which isn't what we want when
  there are validation errors: instead we want the instance of the form
  which was mutated by `.valid?` so that we can display the errors from
  it

dgmstuart committed:
(refactor) Remove 'multiple_choice' conditionals

We have now effectively redefined questions to mean multiple choice
questions, so this functionality is not used.

- this brings the questions controller within an acceptable ABC
  complexity so we can stop rubocop from ignoring it

dgmstuart committed:
(refactor) Extract a saver class

The previous changes pushed the ABC complexity of this method outside
acceptable limits. Extracting the information of exactly what is saved
to the session pushes this down again.

dgmstuart committed:
(feature) Reject partials which don't exist

It's tenuous, but we don't want an attacker to be able to render
arbitrary partials in the page. We can use a method provided by Rails to
check whether a partial exists in a particular directory.

DescriptionPartialChecker is a thin wrapper around lookup_context to
abstract away it's awkward signature. That third boolean argument
determines whether the method searches for partials or non-partials.

Note that we can't pass DescriptionPartialChecker in as a default
argument to DescribeRepair because lookup_context is initialized by
Rails (it contains all the template paths).

dgmstuart committed:
(refactor) Extract a question validator

Now that the description pages depend on partials (and not other
questions) they're not being checked by the 'validate_questions' step.
This refactoring will make it easier to add a check for missing
partials.

dgmstuart committed:
(feature) Check that description partials exist

Special description pages used to be questions so they were checked by
the existing 'next' check. Now that they're related to partials, we need
to check that those partials exist

dgmstuart committed:
(chore) set a minimum coverage

This PR was the final piece in completing our test coverage: the only
remaining gaps are on files generated by Rails which we're not using
yet. It would be good to encourage this to stay high.

This setting will make RSpec return a failure status code if the code
coverage goes below 98%, with a simplecov message explaining why. This
won't block anything locally but will cause the RSpec checks to fail on
CI if code coverage goes down.

I think it's fine for someone in future to push this number down, but
that should be a deliberate choice, rather than as a result of
forgetting to cover part of the code.

leeky commented on app/controllers/application_controller.rb:
```diff
@@ -2,4 +2,8 @@ class ApplicationController < ActionController::Base
   include Authentication
+
+  def selected_answer_store
+    @selected_answer_store ||= SelectedAnswerStore.new(session)
```
Whilst it does make `selected_answer_store` somewhat of a mystery guest,  it does tidy things up quite a bit

👍

leeky commented on app/presenters/describe_repair.rb:
```diff
@@ -0,0 +1,35 @@
+  private
+
+  def diagnosed?
+    !RepairParams.new(@answers).sor_code.nil?
```
How about this instead:

`RepairParams.new(@answers).sor_code.present?`

dgmstuart replied:
That is indeed clearer. I thought this wouldn't be available in this class, but I guess ActiveModel includes it. 

Since we're asking this question of RepairParams in two places, I've created a `diagnosed?` method: that also makes us more flexible if we wanted to redefine what it means to be 'diagnosed` in future.

leeky commented on spec/fixtures/missing_partials.yml:
```diff
@@ -0,0 +1,7 @@
+---
```
I think this file belongs in the next commit(?)

dgmstuart replied:
OOPS!!!!

dgmstuart replied:
Fixed

leeky commented on spec/rails_helper.rb:
```diff
@@ -2,6 +2,7 @@
 SimpleCov.start 'rails' do
   add_group 'Queries', 'app/queries'
   add_group 'Validators', 'app/validators'
+  minimum_coverage 98
```
Should we add `add_group 'Presenters', 'app/presenters'` for completeness?

dgmstuart replied:
Yes! good point

leeky commented on app/controllers/describe_unknown_repair_controller.rb:
```diff
@@ -1,18 +1,25 @@
 class DescribeUnknownRepairController < ApplicationController
   def index
+    @back = Back.new(controller_name: 'questions/start')
```
As discussed, this `Back` shouldn't be here.

dgmstuart replied:
I have eliminated it from history like a time travel B-movie

leeky commented on app/controllers/describe_unknown_repair_controller.rb:
```diff
@@ -1,18 +1,25 @@
   def submit
-    @form = DescribeUnknownRepairForm.new(describe_unknown_repair_form_params)
-    return render :index unless @form.valid?
+    @back = Back.new(controller_name: 'questions/start')
```
And this one too.

leeky reviewed (changes requested):
Just a few minor comments, otherwise nice work :tada:

leeky commented on app/presenters/describe_repair.rb:
```diff
@@ -0,0 +1,38 @@
+    @partial_checker = partial_checker
+  end
+
+  def partial
```
Just playing with this through the frontend, and I was wondering whether this should be...

```ruby
  def partial
    return @details if @partial_checker.exists?(@details)

    if diagnosed?
      'anything_else'
    else
      'describe_problem'
    end
  end
```

i.e. Always use the partial referenced in the `desc` key, but failover to `anything_else` or `describe_problem` depending on whether we have a SOR code or not(?)

I spotted this as I inadvertently had a `sor_code` stored in my session, so it behaved as if the `?details` parameters was being ignored, as `diagnosed?` was always true.

dgmstuart replied:
Agreed - good fix. The problem will still be that the form should be required, but will be optional, but that's a different problem - fixed by resetting the session.

dgmstuart committed:
(refactor) explicitly define a 'diagnosed?' method

Say what we mean when we're checking for the presence of an SOR code.
This also makes us more flexible if we wanted to redefine what it means
to be 'diagnosed` in future.

dgmstuart committed:
(chore) simplecov categories: Services, Presenters

Otherwise these show up in 'ungrouped'

leeky commented on app/controllers/addresses_controller.rb:
```diff
@@ -13,7 +13,7 @@ def create
       return render 'address_searches/create'
     end
 
-    if RepairParams.new(selected_answer_store.selected_answers).sor_code
```
Nice addition :+1:

dgmstuart committed:
(feature) Special descs for diagnosed repairs

Allow diagnosed repairs to have describe repair pages with special text
as well: even when something has been diagnosed, there might still be
some specific information which (though not required) could be useful.

dgmstuart commented:
@leeky ready for another look

leeky commented:
LGTM! :shipit:
